SentenceVectors.ecl

```diff
@@ -0,0 +1,651 @@
+    batchSizeAdj := IF(batchSize = 0, batchSizeCalc, batchSize);
+    // Set up the neural network and do stochastic gradient descent to train
+    // the network.
+    nn := int.SGD(nnShape, trainToLoss, numEpochs, batchSizeAdj, learningRate, negSamples, noProgressEpochs);
+    finalWeights := nn.Train_Dupl(trainDat);
+    // Now extract the final weights for the first layer as the word vectors
+    wVecs := computeWordVectors(finalWeights, vocabulary, nnShape);
+    // And produce the word portion of the model.
+    wMod := makeWordModel(wVecs);
+    // Calculate a vector for each sentence in the corpus to produce the sentence portion
+    // of the model
+    sVecs := sent2vector(wMod, corp.Sentences, vecLen, mapMissingWords := FALSE);
+    sMod := makeSentModel(sVecs);
+    // Concatenate the two portions, unless saveSentences is FALSE.
+    mod := IF(saveSentences, wMod + sMod, wMod);
+    RETURN mod;
+  END;
+  /**
+    * Return the parameters under which the model was generated.  This function
+    * should only be called during the training phase, or when the module was
+    * invoked with the same parameters used in the training phase.  It may return
+    * incorrect information if called with a previously persisted model when the
+    * module was invoked e.g. with default values.
+    * @param mod A model as previously returned from GetModel.
+    * @return A DATASET(Types.TrainStats) containing a single record.
+		* @see Types.TrainStats
+    */
+  EXPORT GetTrainStats(DATASET(TextMod) mod) := FUNCTION
+    vocabSize := COUNT(mod(typ=t_ModRecType.word));
+    nSentences := COUNT(mod(typ=t_ModRecType.sentence));
+    shape := [vocabSize, vecLen, vocabSize];
+    w := int.Weights(shape);  // Module to manage weights
+    batchSizeCalc := (UNSIGNED4)(calConst * vocabSize) / ((1 + negSamples) * nNodes);
+    batchSizeAdj := IF(batchSize = 0, batchSizeCalc, batchSize);
+    TrainStats getStats(TrainStats t) := TRANSFORM
+      SELF.vecLen := vecLen;
+      SELF.nWeights := w.nWeights;
+      SELF.nSlices := w.nSlices;
+      SELF.sliceSize := w.sliceSize;
+      SELF.nWords :=vocabSize;
+      SELF.nSentences := nSentences; 
+      SELF.maxNGramSize := wordNGrams;
+      SELF.nEpochs := numEpochs;
+      SELF.negSamples := negSamples;
+      SELF.batchSize := batchSizeAdj;
+      SELF.minOccurs := minOccurs;
+      SELF.maxTextDist := maxTextDistance;
+      SELF.maxNumDist := maxNumDistance;
+      SELF.discardThreshold := discardThreshold;
+      SELF.learningRate := learningRate;
+      // Updates per batch per node
+      SELF.upbPerNode := SELF.batchSize * (1 + negSamples) * 2 * SELF.vecLen;
+      // Updates per batch
+      SELF.upb := SELF.upbPerNode * nNodes;
+      SELF.updateDensity := SELF.upb / SELF.nWeights;
+      SELF.udPerNode := SELF.upbPerNode / SELF.nWeights;
+    END;
+    stats := DATASET([{0}], TrainStats);
+    statsOut := PROJECT(stats, getStats(LEFT));
```

Comment:
Simpler is DATASET([getStats()]);

A couple of minor notes.  You don't need to pass LEFT to a function if none of the values from left are used.  Also there are generally short-hand ways of creating datasets like this one and DATASET(<n>, transform(counter)).

Reply:
Good to know regarding LEFT.